src/internal/storage/fileset/validator.go

```diff
@@ -0,0 +1,66 @@
+package fileset
+
+import (
+	"context"
+	"fmt"
+	"strings"
+
+	"github.com/pachyderm/pachyderm/v2/src/internal/errors"
+	"github.com/pachyderm/pachyderm/v2/src/internal/pctx"
+	"github.com/pachyderm/pachyderm/v2/src/internal/storage/fileset/index"
+)
+
+var _ FileSet = &pathValidator{}
+
+type pathValidator struct {
+	fs FileSet
+}
+
+// NewPathValidator checks if there is collision between file and dir names.
+func NewPathValidator(fs FileSet) FileSet {
+	pf := &pathValidator{
+		fs: fs,
+	}
+	return pf
+}
+
+func (pf *pathValidator) Iterate(ctx context.Context, cb func(File) error, opts ...index.Option) error {
+	ctx = pctx.Child(ctx, "pathIterator")
+	var prev *index.Index
+	var retErr error = nil
+	if err := pf.fs.Iterate(ctx, func(f File) error {
+		idx := f.Index()
+		if retErr == nil {
```

Comment:
We need to change this logic a bit. We want to return as soon as an error is encountered. The original logic needs to proceed even when an error is encountered because it is also computing the size of the files. So, we should just immediately return here if `CheckIndex` returns an error.

Reply:
Got it